server/manager/src/main/java/org/apache/accumulo/manager/upgrade/UpgradeProgressTracker.java

```diff
@@ -0,0 +1,211 @@
+          AccumuloDataVersion.get());
+      checkArgument(newVersion > rootVersion,
+          "New Root version (%s) must be greater than current Root version (%s)", newVersion,
+          rootVersion);
+      checkArgument(newVersion <= zooKeeperVersion,
+          "New Root version (%s) expected to be <= ZooKeeper version (%s)", newVersion,
+          zooKeeperVersion);
+      checkArgument(newVersion > metadataVersion,
+          "New Root version (%s) must be greater than current Metadata version (%s)", newVersion,
+          metadataVersion);
+      rootVersion = newVersion;
+      put(context, this);
+    }
+
+    public int getMetadataVersion() {
+      return metadataVersion;
+    }
+
+    public synchronized void updateMetadataVersion(ServerContext context, int newVersion)
+        throws KeeperException, InterruptedException {
+      requireNonNull(context, "ServerContext must be supplied");
+      checkArgument(newVersion <= AccumuloDataVersion.get(),
+          "New version (%s) cannot be larger than current data version (%s)", newVersion,
+          AccumuloDataVersion.get());
+      checkArgument(newVersion > metadataVersion,
+          "New Metadata version (%s) must be greater than current version (%s)", newVersion,
+          metadataVersion);
+      checkArgument(newVersion <= zooKeeperVersion,
+          "New Metadata version (%s) expected to be <= ZooKeeper version (%s)", newVersion,
+          zooKeeperVersion);
+      checkArgument(newVersion <= rootVersion,
+          "New Metadata version (%s) expected to be <= Root version (%s)", newVersion, rootVersion);
+      metadataVersion = newVersion;
+      put(context, this);
+    }
+  }
+
+  private static String getZPath(ServerContext context) {
+    return context.getZooKeeperRoot() + Constants.ZUPGRADE_PROGRESS;
+  }
+
+  public static boolean upgradeInProgress(ServerContext context)
+      throws KeeperException, InterruptedException {
+    requireNonNull(context, "ServerContext must be supplied");
+    final String zpath = getZPath(context);
+    final ZooSession zs = context.getZooSession();
+    return zs.exists(zpath, null) != null;
+  }
+
+  private static void put(ServerContext context, UpgradeProgress cv)
+      throws KeeperException, InterruptedException {
+    requireNonNull(context, "ServerContext must be supplied");
+    requireNonNull(cv, "ComponentVersions object  must be supplied");
+    final String zpath = getZPath(context);
+    final ZooSession zs = context.getZooSession();
+    Stat stat = zs.exists(zpath, null);
+    if (stat == null) {
+      zs.create(zpath, GSON.get().toJson(cv).getBytes(UTF_8), ZooUtil.PUBLIC,
+          CreateMode.PERSISTENT);
+      cv.znodeVersion = zs.exists(zpath, null).getVersion();
+    } else {
+      try {
+        zs.setData(zpath, GSON.get().toJson(cv).getBytes(UTF_8), cv.znodeVersion);
+      } catch (KeeperException e) {
+        if (e.code() == KeeperException.Code.BADVERSION) {
```

Comment:
Each separate code has it's own KeeperException subclass. You could catch the one for KeeperException.BadVersionException directly in an earlier catch clause, so you don't have to do this check of the code.